File: examples/helloworld/helloworld.py
import sys
import time
import json
from mpi4py import MPI

import llmEngine

def main():
  comm = MPI.COMM_WORLD
  rank = comm.Get_rank()
  size = comm.Get_size()

  # Creating LLM Engine object
  llmEngine_ = llmEngine.LLMEngine()

  def fc(task):
    print("Hello World")

    # This instance can terminate the llmEngine
    llmEngine_.terminate()

  llmEngine_.registerFunction("Hello World", fc)

  # Initializing LLM engine
  isRoot = llmEngine_.initialize(sys.argv)
  
  # Let only the root instance deploy the LLM engine
  if isRoot:

    # Checking arguments
    if len(sys.argv) != 2:
      print("Error: Must provide the request file as argument.", file=sys.stderr)
      llmEngine_.abort()
      sys.exit(1)

    configFilePath = str(sys.argv[1])

    # Parsing request file contents to a JSON object
    with open(configFilePath, 'r') as file:
      configJs = json.load(file)

    # Running LLM Engine
    llmEngine_.run(configJs)

  # Finalizing LLM engine
  llmEngine_.finalize()


if __name__ == "__main__":
  main()

File: examples/helloworld/helloworld.cpp
#include <stdio.h>
#include <fstream>
#include "llm-engine/llm-engine.hpp"

int main(int argc, char *argv[])
{
  // Creating LLM Engine object
  llmEngine::LLMEngine llmEngine;

  // Listen request function -- it expects an outside input and creates a request
  std::string requestOutput;
  llmEngine.registerFunction("Hello World", [&](llmEngine::Task *task) {
    printf("Hello World\n");

    // This instance can terminate the llmEngine
    llmEngine.terminate();
  });

  // Initializing LLM engine
  auto isRoot = llmEngine.initialize(&argc, &argv);

  // Let only the root instance deploy the LLM engine
  if (isRoot)
  {
    // Checking arguments
    if (argc != 2)
    {
      fprintf(stderr, "Error: Must provide the request file as argument.\n");
      llmEngine.abort();
      return -1;
    }

    // Getting config file name from arguments
    std::string configFilePath = std::string(argv[1]);

    // Parsing request file contents to a JSON object
    std::ifstream ifs(configFilePath);
    auto          configJs = nlohmann::json::parse(ifs);

    // Running LLM Engine
    llmEngine.run(configJs);
  }

  // Finalizing LLM engine
  llmEngine.finalize();
}